hci.c

```diff
@@ -0,0 +1,217 @@
+{
+    manager->prefix = strdup(prefix);
+    manager->timer_begin = hci_now(manager);
+    manager->timer_query_begin = manager->timer_begin;
+
+    manager->WallClockTimeLimit = WallClockTimeLimit;
+    manager->AutoCheckPointTime = AutoCheckPointTime;
+    manager->LongestTimeBetweenQueries = 0;
+}
+
+/* override the result of hci_now; for unit testing -- we can't rely on MPI_Wtime there! 
+ * this function can be called before hci_init. */
+int
+hci_override_now(HCIManager * manager, double now)
+{
+    manager->_now = now;
+    manager->OVERRIDE_NOW = 1;
+}
+
+static double
+hci_get_elapsed_time(HCIManager * manager)
+{
+    double e = hci_now(manager) - manager->timer_begin;
+    return e;
+}
+
+static
+void hci_update_query_timer(HCIManager * manager)
+{
+    double e = hci_now(manager);
+    e = e - manager->timer_query_begin;
+    if(e > manager->LongestTimeBetweenQueries)
+        manager->LongestTimeBetweenQueries = e;
+
+    manager->timer_query_begin = e;
+}
+
+/*
+ * query the filesystem for HCI commands;
+ * returns the content of the file or NULL; collectively
+ * */
+static char *
+hci_query_filesystem(HCIManager * manager, char * filename)
+{
+    int ThisTask;
+    int NTask;
+    MPI_Comm_size(MPI_COMM_WORLD, &NTask);
+    MPI_Comm_rank(MPI_COMM_WORLD, &ThisTask);
+    int size = 0;
+    char * content = NULL;
+    if(ThisTask == 0) {
+        char * fullname = fastpm_strdup_printf("%s/%s", manager->prefix, filename);
+        content = fastpm_file_get_content(fullname);
+        if(content) {
+            size = strlen(content);
+            remove(fullname);
+        } else {
+            size = -1;
+        }
+        free(fullname);
+    }
+    MPI_Bcast(&size, 1, MPI_INT, 0, MPI_COMM_WORLD);
+
+    if(size != -1) {
+        if(ThisTask != 0) {
+            content = calloc(size + 1, 1);
+        }
+        MPI_Bcast(content, size+1, MPI_BYTE, 0, MPI_COMM_WORLD);
+    } else {
+        content == NULL;
+    }
+
+    return content;
+}
+
+static char *
+hci_query_timeout(HCIManager * manager)
+{
+    double now = hci_get_elapsed_time(manager);
+    /*
+     * factor 0.9 is a safety tolerance
+     * for possible inconsistency between measured time and the true wallclock
+     *
+     * If there likely isn't time for a new query, then we shall timeout as well.
+     * */
+
+    if (now + manager->LongestTimeBetweenQueries < manager->WallClockTimeLimit * 0.9) {
+        return NULL;
+    }
+
+    /* any freeable string would work. */
+    return calloc(32, 1);
+}
+
+static char *
+hci_query_auto_checkpoint(HCIManager * manager)
+{
+    /*How long since the last checkpoint?*/
+    if(manager->AutoCheckPointTime <= 0) return NULL;
+
+    double now = hci_get_elapsed_time(manager);
+    if(now - manager->TimeLastCheckPoint >= manager->AutoCheckPointTime) {
+        /* any freeable string would work */
+        return calloc(32, 1);
```

Comment:
These two functions could return an int

Reply:
I wanted them to have the same interface -- int would break that.

Reply:
In that case I would suggest having all functions return int, 1 == do something, 0 == do nothing, and hci_query_filesystem take an optional char * file_contents argument to store the file contents.

Reply:
Yes. I'll do that. Is it has to be a `char**`, though.